File: src/Pieces/Bishop.java
package Pieces;

import Board.Board;
import Utils.Movement;

import java.util.ArrayList;
import java.util.List;

public class Bishop extends Piece {

    public Bishop(boolean side, int value, String representation) {
        super(side, value, representation);
    }

    private List<int[]> getAllBishopMovesA(int[] move) {
        Movement movementUtil = new Movement();
        List<int[]> allMovements = new ArrayList<>();
        int x = move[0];
        int y = move[1];
        while (movementUtil.inBounds(x, y)) {
            x--;
            y--;
            if (!allMovements.contains(new int[] {x, y})) { allMovements.add(new int[] {x, y}); }
        }
        return allMovements;
    }

    private List<int[]> getAllBishopMovesB(int[] move) {
        Movement movementUtil = new Movement();
        List<int[]> allMovements = new ArrayList<>();
        int x = move[0];
        int y = move[1];
        while (movementUtil.inBounds(x, y)) {
            x++;
            y++;
            if (!allMovements.contains(new int[] {x, y})) { allMovements.add(new int[] {x, y}); }
        }
        return allMovements;
    }
    private List<int[]> getAllBishopMovesC(int[] move) {
        Movement movementUtil = new Movement();
        List<int[]> allMovements = new ArrayList<>();
        int x = move[0];
        int y = move[1];
        while (movementUtil.inBounds(x, y)) {
            x++;
            y--;
            if (!allMovements.contains(new int[] {x, y})) { allMovements.add(new int[] {x, y}); }
        }
        return allMovements;
    }

    private List<int[]> getAllBishopMovesD(int[] move) {
        Movement movementUtil = new Movement();
        List<int[]> allMovements = new ArrayList<>();
        int x = move[0];
        int y = move[1];
        while (movementUtil.inBounds(x, y)) {
            x--;
            y++;
            if (!allMovements.contains(new int[] {x, y})) { allMovements.add(new int[] {x, y}); }
        }
        return allMovements;
    }

    @Override
    public boolean checkMove(Board board, int[] move, boolean side) {
        Movement movementUtil = new Movement();
        List<int[]> allowedMovements = new ArrayList<>();
        int x, y;

        List<List<int[]>> allMovements = new ArrayList<>(4);
        allMovements.add(0, getAllBishopMovesA(move));
        allMovements.add(1, getAllBishopMovesB(move));
        allMovements.add(2, getAllBishopMovesC(move));
        allMovements.add(3, getAllBishopMovesD(move));

        for (List<int[]> ints : allMovements) {
            for (int[] movement : ints) {
                x = movement[0];
                y = movement[1];
                if (movementUtil.inBounds(x, y)) {
                    if (board.getFigure(x, y).getClass().isInstance(new Empty())) {
                        allowedMovements.add(new int[] {x, y});
                    } else {
                        if (board.getFigure(x, y).getSide() != board.getFigure(move[0], move[1]).getSide()) {
                            allowedMovements.add(new int[] {x, y});
                        }
                        break;
                    }
                }
            }
        }

        return movementUtil.isAllowed(allowedMovements, move);
    }

}

File: Python/src/Pieces/Bishop.py
from Pieces.Piece import Piece, Empty
from Utils.Movement import Movement

class Bishop(Piece):
    def __init__(self, side, value, representation):
        self.side = side 
        self.value = value 
        self.representation = representation
    
    def __str__(self): 
        return self.representation

    def getAllBishopMovesA(self, move): 
        movement = Movement()
        allMovements = []
        while (movement.inBounds(move[0], move[1])):
            move[0] -= 1 
            if not [move[0], move[1]] in allMovements: 
                allMovements.append([move[0], move[1]])
        return allMovements

    def getAllBishopMovesB(self, move): 
        movement = Movement()
        allMovements = []
        while (movement.inBounds(move[0], move[1])):
            move[0] += 1 
            if not [move[0], move[1]] in allMovements: 
                allMovements.append([move[0], move[1]])
        return allMovements

    def getAllBishopMovesC(self, move): 
        movement = Movement()
        allMovements = []
        while (movement.inBounds(move[0], move[1])):
            move[1] -= 1 
            if not [move[0], move[1]] in allMovements: 
                allMovements.append([move[0], move[1]])
        return allMovements

    def getAllBishopMovesD(self, move): 
        movement = Movement()
        allMovements = []
        while (movement.inBounds(move[0], move[1])):
            move[1] += 1 
            if not [move[0], move[1]] in allMovements: 
                allMovements.append([move[0], move[1]])
        return allMovements

    def checkMove(self, board, move, side):
        movementUtil = Movement()
        allowedMovements = []
        allMovements = [self.getAllBishopMovesA(move), self.getAllBishopMovesB(move), 
            self.getAllBishopMovesC(move), self.getAllBishopMovesD(move)]
        for ints in allMovements: 
            for movement in ints: 
                x = movement[0]
                y = movement[1]
                if movementUtil.inBounds(x, y): 
                    if type(board.getFigure(x, y)) is Empty("o").__class__:
                        if movementUtil.isFree(board, x, y, side, board.getFigure(move[0], move[1])):
                            allowedMovements.append([x, y])
                    else: 
                        if board.getFigure(x, y).side != board.getFigure(move[0], move[1]).side: 
                            allowedMovements.append([x, y])
                        break
        return movementUtil.isAllowed(allowedMovements, move)
